feat: add custom key labels in cfg

jtroo opened the pull request:
This commit adds a new configuration item `defcustomkeys`. This allows a user to define custom keys that can be used without aliases in `defsrc`, `deflayer` and other config items.

The documentation is updated to hopefully encourage community-created resources for non-US keyboard `defsrc` configurations.

Related to #204

jtroo commented on src/cfg/mod.rs:
```diff
@@ -458,6 +474,34 @@ fn parse_defcfg(expr: &[SExpr]) -> Result<HashMap<String, String>> {
+            Some(v) => v.atom()
+                .ok_or_else(|| anyhow!("defcustomkeys contains a list; no lists are allowed"))
+                .and_then(|osc| osc.parse::<u16>().map_err(|_| anyhow!("defcustomkeys unknown number {osc}")))
+                .and_then(|osc| OsCode::from_u16(osc).ok_or_else(|| anyhow!("defcustomkeys unknown number {osc}")))?,
```
I'm pretty sure this is incorrect; this should use `as`, since `from_u16` will differ between Linux and Windows. Probably good to add a unit test for this - make sure mapping between Linux and Windows is the same

gerhard-h commented:
first compiled main then switched to branch 204-locale-specific-defsrc

```
;;  ü 186 ;; 0xba
(defcustomkeys
 ü 186
)
```
ü is also added to defsrc.

when starting kanata I get:

 [DEBUG] (1) kanata::cfg: custom mapping: ü 127

a different number ?!

I can't remap ü and  `(macro ü ) ` outputs nothing

jtroo commented:
Thanks for trying it out; I'll do some more investigation... thought I had it right 😅

jtroo commented:
I realized that the mapping will have to differ in Windows hooks vs (Linux/Windows-Interception) due to https://github.com/jtroo/kanata/issues/152 😢

jtroo commented:
@gerhard-h I've updated the code so I think it should work now, and I've added a sample config: https://github.com/jtroo/kanata/pull/215/files#diff-456a6bc32d7cb61b8307053275c52beb1e15e364898cae037cdb59737b77eaf7R13

Please let me know if that works when you get a chance. I updated the documentation to point to an as-of-yet unreleased version of the keyboard tester binaries; I'll be updating those next.

gerhard-h commented:
cargo build **--release**
   Compiling kanata v1.1.0 (C:\path\kanata)
error: linking with `link.exe` failed: exit code: 1104
  |
  = note: "C:\\Program Files\\Microsoft Visual Studio\\2022\\Professional\\VC\\Tools\\MSVC\\14.32.31326\\bin\\HostX64\\x64\\link.exe"...

I get this since today, it is not related to branch 204-locale-specific-defsrc

In general I would blame Windows Update, but have not even rebooted and rust/cargo has this suspicious autoupdate feature even for each compile. 

Do you use rust inside a dev container?

if I omitt --release it compiles.

sofar I just tested ü ... worked, It will take me some time to include and thourougly test the other keys.

jtroo commented:
Yea seems like it's a Windows/Rust environment issue.

In Windows I develop and compile on the host; I have a Windows 11 machine and a Windows 10 machine, haven't found an issue yet.

gerhard-h commented:
`[ERROR] live reload failed + is a default key name in kanata; it cannot be used in defcustomkeys`

In the long run it would be great to have a config option to get rid of the US mappings because naming `+` `+*~` but not naming `ß` `ß?\` is a little bit confusing for beginners.

gerhard-h commented:
I used this defcustomkeys - I would prefer deflocalkeys or definternationalkeys - and found no errors
I can not think of a case any german would  want a different definition (besides being on a different plattform) 
```
(defcustomkeys
  ü    186
  +    187
  #    191
  ö    192
  ß    219
  ^    220
  ´     221
  ä    222
  <    226
)
```
I also defined aliases for all modified characters 
* but using unicode look-a-likes is maybe not for everyone
* also getting rid of dead keys will not please people who dare to write a lot é, â, ò,...
```
(defalias
  ;; shifted german keys
  ! S-1
  ˝ S-2  ;; unicode 02DD ˝ look-a-like is used because @" is no valid alias, to be displayed correctly
         ;; in console requires a font that can - e.g. cascadia
  §	S-3
  $	S-4
  %	S-5
  &	S-6
  /	S-7
  ﴾	S-8  ;; unicode FD3E ﴾ look-a-like is used because @( is no valid alias, to be displayed correctly...
  ﴿	S-9  ;; unicode FD3F ﴿ look-a-like is used because @) is no valid alias, to be displayed correctly ...
  =	S-0
  ? S-ß
  * S-+
  ' S-#
  ; S-,
  : S-.
  _ S--
  > S-<
  < <   ;; not really needed but having @< and @> looks consistent

  ;; change dead keys in normal keys
  ´ (macro ´ spc )	  ;; ´ 
  ` (macro S-´ spc )  ;; `
  ^ (macro ^ spc )    ;; ^ = \ - shifting @^ will produce an incorrect space now
  ° S-^
  
  ;; AltGr german keys
  ~ A-C-+
  \ A-C-ß
  ẞ A-C-S-ß
  | A-C-<
  } A-C-0
  { A-C-7
  ] A-C-9
  [ A-C-8	
  € A-C-e
  @ A-C-q
  ² A-C-2
  ³ A-C-3
  µ A-C-m
)
```
I think including a default defcustomskeys and a defalias section with "shifted" keys in the tutorial kbd
and adding a comment where to find the localized files may be a good starting point.

Does kanata know on wich plattform it is currently running? that way 
defcustomkeys-win, defcustomkeys-mac, defcustomkeys-nix could all be active at the same time (for tutorial.kbd to be functional from the get go)

gerhard-h commented:
> Yea seems like it's a Windows/Rust environment issue.

Ihave not even restarted vscode and today --release compiles again ... magic
